### ISSUE TITLE
Listing labels only returns the first page of items

### ISSUE THREAD

**Issue by sorenmortensen** [softprops/hubcaps#109]
I'm attempting to list the labels in a repository using `hubcaps`, and finding that for repositories that have more than one page of labels (e.g. [rust-lang/cargo](https://www.github.com/rust-lang/cargo/labels)), I'm only getting the first page of results (i.e. not all the labels).

I found [this page on pagination](https://developer.github.com/v3/guides/traversing-with-pagination/) from the API reference, and I'm not sure if this isn't implemented in `hubcaps` or if I'm just missing something.

This type of thing should reproduce the issue. Labels are returned, but only the first page of them.

```rust
let mut core = Core::new().expect("Couldn't initialise core");

let github = Github::new(
    concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION")),
    Some(Credentials::Token(
        "<token>".to_owned(),
    )),
    &core.handle(),
);

let labels = core.run(
    github
        .repo("rust-lang".to_owned(), "cargo".to_owned())
        .labels()
        .list(),
);

labels
    .expect("No labels")
    .iter().for_each(|l| println!("{}", l.name));
```

My implementation that isn't working is at sorenmortensen/ghtool@0de90b9233c203243464c411cc99aa54d0a81f6e. Feel free to clone it and try with:

```bash
git clone git@github.com:sorenmortensen/ghtool.git
git checkout label-list
cargo run -- -T "<access-token>" label list rust-lang/cargo
```

**Comment by softprops** [comment]
sounds like a bug, I'll take a look

**Comment by softprops** [comment]
actually not a bug just a missing feature. For things that return lists you'll typically find two methods, one call list which returns a future returning the first page of results for the simple cases, and one called iter() which provides a stream over all of the items (lazily fetched) abstracting over pages over results. I don't think there's an impl for repo labels - http://lessis.me/hubcaps/hubcaps/labels/struct.Labels.html. I'll take that as a feature request.

**Comment by sorenmortensen** [comment]
Would you like me to file a separate issue for the feature request?

**Comment by softprops** [comment]
Nah this is fine. I'll try to take a look at it this weekend

**Comment by sorenmortensen** [comment]
I took a crack at implementing this by following the examples of other types that implement `iter()` (sorenmortensen/hubcaps@025fe91). Unfortunately, I'm now seeing a different bug with not only `Labels.iter()` but also other implementations (every one that I was able to test exhibited this behaviour) such as `UserRepositories.iter()`. Instead of getting all of the items across each of the pages, I seem to be getting infinitely repeating copies of the first page.

If I list the repositories owned by [rust-lang](https://www.github.com/rust-lang) using the following code...

```rust
let mut core = Core::new().expect("No core");

let github = Github::new(
    concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION")),
    Some(Credentials::Token("<token>".to_owned())),
    &core.handle(),
);

let count = Arc::new(Mutex::new(0_usize));

core.run(
    github
        .user_repos("rust-lang".to_owned())
        .iter(&UserRepoListOptions::default())
        .for_each(move |repo| {
            let count_clone = count.clone();
            let mut count_guard = count_clone.lock().unwrap();
            *count_guard += 1;

            println!("{count}: {name}", count = *count_guard, name = repo.name);

            Ok(())
        })
).expect("Error listing repos");
```

...then I get the following output:

```
1: blog.rust-lang.org
2: book
3: cargo
4: compiler-rt
5: crates.io
6: crates.io-index
7: gedit-config
8: hoedown
9: jemalloc
10: kate-config
11: libc
12: lld
13: llvm
14: meeting-minutes
15: nano-config
16: regex
17: rfcs
18: rust
19: rust-by-example
20: rust-central-station
21: rust-enhanced
22: rust-guidelines
23: rust-installer
24: rust-mode
25: rust-playpen
26: rust-roadmap
27: rust-wiki-backup
28: rust-www
29: rust.vim
30: subteams
31: subteams
32: blog.rust-lang.org
33: book
34: cargo
35: compiler-rt
36: crates.io
37: crates.io-index
38: gedit-config
39: hoedown
40: jemalloc
41: kate-config
42: libc
43: lld
44: llvm
45: meeting-minutes
46: nano-config
47: regex
48: rfcs
49: rust
50: rust-by-example
51: rust-central-station
52: rust-enhanced
53: rust-guidelines
54: rust-installer
55: rust-mode
56: rust-playpen
57: rust-roadmap
58: rust-wiki-backup
59: rust-www
60: rust.vim
61: subteams
62: blog.rust-lang.org
63: book
64: cargo
65: compiler-rt
66: crates.io
67: crates.io-index
68: gedit-config
69: hoedown
70: jemalloc
71: kate-config
72: libc
73: lld
74: llvm
75: meeting-minutes
76: nano-config
77: regex
78: rfcs
79: rust
80: rust-by-example
81: rust-central-station
82: rust-enhanced
83: rust-guidelines
84: rust-installer
85: rust-mode
86: rust-playpen
87: rust-roadmap
88: rust-wiki-backup
89: rust-www
90: rust.vim
```

It looks to me as though the same repositories are repeating over and over again, with a period of approx. 30 (which I believe is the length of a page). I let this run up to at least `330`, and it continued to repeat. [rust-lang](https://www.github.com/rust-lang) only has 31 repositories.

Now, I leave open the possibility that I'm doing something stupid and consuming the `Stream` in entirely the wrong way (I'm quite new to futures). But as far as I can tell I followed the examples (e.g.
 the [repos example](https://github.com/softprops/hubcaps/blob/master/examples/repos.rs)) pretty closely, and I'm not getting the results I'd expect.

**Comment by sorenmortensen** [comment]
Update: I just ran the exact code from [the repos example](https://github.com/softprops/hubcaps/blob/master/examples/repos.rs) and it exhibits the same behaviour.

**Comment by softprops** [comment]
I just noticed ( and merged ) a fix to the a more general bug with the iter methods as well as added a new iter method on labels

I added a new example app [here](https://github.com/softprops/hubcaps/blob/6d059fdaf77add294ebd58093ed462ed80d1d9db/examples/labels.rs#L29-L33) for labels

```bash
cargo run --example labels
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/examples/labels`
A-build-dependencies
A-build-execution
A-build-plan
A-build-scripts
A-building-cargo-itself
A-caching
A-cargo-internals
A-configuration
A-console-output
A-crate-dependencies
A-cross-compiling
A-debugging
A-dependency-resolution
A-dev-dependencies
A-documenting-cargo-itself
A-environment-variables
A-errors
A-examples
A-external-dependencies
A-features
A-git
A-interacts-with-crates.io
A-linkage
A-manifest
A-new-subcommand
A-optional-dependencies
A-overrides
A-patch
A-profiles
A-rebuild-detection
Command-publish
A-replace
A-source-replacement
A-testing-cargo-itself
A-tooling
A-vcs
A-workspaces
B-scheduled-change
C-bug
C-cleanup
C-enhancement
C-feature-request
C-question
Command-bench
Command-build
Command-check
Command-clean
Command-doc
Command-fetch
Command-generate-lockfile
Command-git-checkout
Command-init
Command-install
Command-locate-project
Command-login
Command-metadata
Command-new
Command-owner
Command-package
Command-pkgid
relnotes
Command-read-manifest
Command-run
Command-rustc
Command-rustdoc
Command-search
Command-test
Command-uninstall
Command-update
Command-verify-project
Command-version
Command-yank
E-easy
E-hard
E-help-wanted
E-mentor
E-needs-mentor
Feature accepted
I-nominated
Needs design
O-linux
O-macos
O-windows
P-high
T-cargo
WG-cargo-registries
blocked
```

**Comment by softprops** [comment]
published a new version, 0.4.4
